text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+The process of restarting the network could be very similar to the process of starting it. In its essence, it is just like starting the network from scratch, but with a few differences:
+
+- There is a different set of genesis validators, which is based on the bonded validators of the last correct block, with the equivocators removed.
+- There is a different initial global state: it is not empty anymore, but it is the post-state of the last correct block.
+- Other details like the protocol version can differ as well, but they are not of the main concern for the restart process itself.
+
+The information necessary for restarting the network can (and probably should) be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Here we will just summarize the exact steps to be performed when an equivocation catastrophe happens.
+
+1. An equivocation catastrophe is detected. Once a single validator detects it, we can expect all the validators to detect it eventually as well (even if just due to them receiving the evidence from the validator that detected it first).
+2. Consensus gets suspended, though validators continue to gossip. The users (validator nodes' owners) receive notice about the catastrophe and start the process of social consensus.
+3. The users agree about the last correct block, the transactions that need to be replayed on top of it and the set of equivocating validators (exchanging evidence if necessary).
+4. A chainspec for the restart is manually constructed and agreed upon. This includes:
+    - the initial set of validators for the restarted network which excludes the equivocators,
+    - the initial global state hash for the restarted network
+5. The network is being restarted using the chainspec constructed above.
+
+## Drawbacks
+
+[drawbacks]: #drawbacks
+
+The single drawback is that this proposal requires manual intervention in the case of an equivocation catastrophe. An ideal situation would be if the network could handle these automatically, but this is most probably impossible.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+As mentioned above, automatic handling of equivocation catastrophes is probably impossible (as it would effectively mean constructing a consensus algorithm that exceeds proven theoretical limitations), so manual intervention is really the single option of dealing with such situations.
+
+An alternative to social consensus would be to have a designated party deal with the situation. However, such a solution would go against the goal of creating a truly distributed system. In this alternative, the party responsible for handling equivocation catastrophes would effectively single-handedly control the network, which is something we want to avoid.
```

Comment:
I totally agree!
But our _recommended_ process could still be to wait for CasperLabs to make a restart proposal (propose the new starting block and validator set), carefully check that, and if it's agreeable, go with it.
Because that's effectively also our default process for protocol updates: CasperLabs proposes them, and hopes that the validators and the community will accept them.

We can argue that that's not centralization, because if the community decided at any point to stop following us, and agreed on a different process for updating the software and handling equivocation catastrophes, they could do that.

The downside for us would be that we'd have to do a good job with those recommendations: We'd need to have someone "on-call" 24 hours/day, who could quickly resolve such situations.

Reply:
@afck, perhaps some consensus protocol issues alluded to with respect to formalizing a process could be circumvented via reliance on an _irrelevant_ third party, e.g., getting people into a Zoom room, where you can't equivocate on a vote without magic hacking skills and/or telepathy?

Reply:
The means of communication is yet another question, yes. I'd recommend several independent communication channels, though; e.g. at least some forum _and_ Zoom, just in case: So that if Zoom is censoring you or even just having technical problems that prevent you from joining, you can tell the others via the forum.

But that's kind of orthogonal to the _process_: What do we _do_ once we're in the Zoom room? I think it's fine to not make any recommendation here. I think it's also fine to kind of promise that we'll make a proposal in such cases.

Reply:
Well, it's not orthogonal, because being in a Zoom meeting (or any other channel where everyone sees everyone simultaneously) means that the points above about theoretical limits to consensus mechanisms no longer apply, since whole classes of misbehaviors are no longer possible (everyone heard you vote simultaneously, there's no equivocating). Of course, then you run into the limits imposed by social choice theory, but it's easier to think about voting rules relative to consensus algorithms.

Reply:
Good point, yes.
So are you suggesting to not recommend a particular process at all (other than having a call with all participants), or to define a precise process that basically implements a "consensus protocol" with perfect synchrony (i.e. live video call)?